### PR TITLE
'dev' branch with basic functionality merge

### LINKED ISSUE #1: Write tests.
https://github.com/bredbrains/tthk-bot/blob/89625daaa09792f3399c19e56b83dbdf99c5237a/tests.py#L4-L6

### LINKED ISSUE #3: 'Changes' command.
**User should have an opportunity to get next information:**
* Changes by date
* Changes by group
* Changes by teacher

### LINKED ISSUE #4: 'Consultations' functionality.
Consultations search by teacher's name.

### LINKED ISSUE #7: Keyboard displaying during message.

### PR DESCRIPTION
**Pull request by blinchk**

### CODE REVIEW

File: models/consultation.py [hunk 1]
```diff
@@ -1,9 +1,9 @@
 class ConsultationTemplate:
     @staticmethod
     def convert(consultation):
-        return f"Учитель {consultation.teacher} \nКабинет: {consultation.room}\nПочта: {consultation.email}" \
+        return f"👨‍🏫 Учитель {consultation.teacher} \nКабинет: {consultation.room}\n✉ Почта: {consultation.email}" \
```
Comment (blinchk) [inline models/consultation.py:4]: Teacher's email not everywhere displayed and written. I recommend to add some if statement for this.

File: api.py [hunk 2]
```diff
@@ -1,3 +1,118 @@
...
+        for item in changes_json:
+            change = Change(item["dayofweek"], item["date"], item["group"], item["lessons"], item["teacher"],
+                            item["room"])
+            if group in change.group:
+                changes.append(change.get_str())
+        if changes:
+            return '\n'.join(changes)
+        return None
+
+    # --------------------------------------------------------------------------------------------------------------------
```
Comment (blinchk) [inline api.py:51]: Comments will be useless, when class names obviously takes thems value.

File: api.py [hunk 3]
```diff
@@ -1,3 +1,118 @@
...
+import json
+
+import requests
+
 class API:
     def __init__(self):
-        self.api_url = "http://tthk-api.herokuapp.com/"
+
+        # --------------------------------------------------------------------------------------------------------------------
+        #                                         Url getting
```
Comment (blinchk) [inline api.py:11]: This url isn't got from somewhere, it already defined by us.

File: api.py [hunk 4]
```diff
@@ -1,3 +1,118 @@
...
+        if consultations:
+            return '\n'.join(consultations)
+        return None
+
+    def search_teacher_by_name(self, name):
+        r = requests.get(self.api_url + 'consultations')
+        consultations_json = eval(json.dumps(r.json()).replace("null", "None"))["data"]
+        consultations = []
+        for item in consultations_json:
+            for temporal in item["times"]:
```
Comment (blinchk) [inline api.py:132]: Better to define external variable with name *time* for this for-loop.

File: api.py [hunk 5]
```diff
@@ -1,3 +1,118 @@
...
+                consultation = Consultation(item["teacher"], item["room"], item["email"], item["department"],
+                                            item["times"], temporal["weekday"], temporal["time"])
+            if name in consultation.teacher:
+                consultations.append("я нашел учителя : " + consultation.get_teachers())
```
Comment (blinchk) [inline api.py:136]: Typo with large letter.

File: bot.py [hunk 6]
```diff
@@ -1,15 +1,184 @@
...
+        #                                        Basic bot and api initialisation
+        # --------------------------------------------------------------------------------------------------------------------
         self.bot = bot.Bot(token)
+        self.api = API()
+
+        # --------------------------------------------------------------------------------------------------------------------
+        #                                           Keyboards initialisation
+        # --------------------------------------------------------------------------------------------------------------------
+        self.keyboard = Keyboard(one_time=False, inline=False) \
+            .add(Text("изменения в расписании"), color=KeyboardButtonColor.POSITIVE) \
```
Comment (blinchk) [inline bot.py:20]: Commands should start with big letter on the start.

File: api.py [hunk 7]
```diff
@@ -1,3 +1,118 @@
+from models.change import Change
+from models.consultation import Consultation
+import json
+
+import requests
+
 class API:
```
Comment (blinchk) [inline api.py:7]: All, what u can commend in this file is:
* This class sends GET-requests to tthkAPI.

And this all as well.

File: api.py [hunk 8]
```diff
@@ -1,3 +1,118 @@
...
+        consultations = []
+        for item in consultations_json:
+            for temporal in item["times"]:
+                consultation = Consultation(item["teacher"], item["room"], item["email"], item["department"],
+                                            item["times"], temporal["weekday"], temporal["time"])
+            if name in consultation.teacher:
+                consultations.append("я нашел учителя : " + consultation.get_teachers())
+        if consultations:
+            return '\n'.join(consultations)
+        return None
```
Comment (blinchk) [inline api.py]: This clones previous methods. DRY.

File: models/consultation.py [hunk 9]
```diff
@@ -1,7 +1,29 @@
...
+
+class TeacherSearchingTemplate:
+    @staticmethod
+    def convert(consultation):
+        return f"{consultation.teacher}"
+
+
 class Consultation:
-    def __init__(self, teacher, room, email, department, times):
+    def __init__(self, teacher, room, email, department, times, weekday, time):
```
Comment (blinchk) [inline models/consultation.py:16]: Better to define *time* array for this object. Like: 1 teacher have 2-5 times of consultations.

File: api.py [hunk 10]
```diff
@@ -1,3 +1,118 @@
...
+            consultations.append(consultation.get_str())
+        if consultations:
+            return '\n'.join(consultations)
+        return None
+
+    # --------------------------------------------------------------------------------------------------------------------
+    #                                         Teacher searching
+    # --------------------------------------------------------------------------------------------------------------------
+
+    def search_teachers(self):
```
Comment (blinchk) [inline api.py:108]: Why teacher searching should be here? And what info that should contain, if rooms and timetable always changes?

File: bot.py [hunk 11]
```diff
@@ -1,15 +1,184 @@
...
+                    await message.answer(search[:900])
+            elif teacher:
+                await message.answer("нет такого учителя, которого вы указали.")
+            else:
+                await message.answer("нет данных об учителях.")
+
+        # --------------------------------------------------------------------------------------------------------------------
+        #                                               Command list
+        # --------------------------------------------------------------------------------------------------------------------
+        @self.bot.on.chat_message(text=["-thk?список команд",
```
Comment (blinchk) [inline bot.py:119]: Command list is useless, when we use keyboard. Keyboard is added to solve this problem.

File: main.py [hunk 12]
```diff
@@ -1,7 +1,7 @@
 from bot import Bot
 import os
 
-token = os.environ["token"]
+token = "28ad13f7d1dacf5542ab985a2e8b4168c2eebc7fdf6a294a9ab8338e7e6c79ce59f3c9cb10a5e67fcceee"
```
Comment (blinchk) [inline main.py:4]: You should use os.environ["token"] to get your token. Mustn't to define it inside .py files.

File: models/change.py [hunk 13]
```diff
@@ -10,8 +10,41 @@ class ChangeTypes(Enum):
 
 
 class ChangeTemplate:
-    def __init__(self, change):
-        self.change = change
+    @staticmethod
+    def convert(change):
+        change.type = change.get_type()
+        change_date = ""
+        if change.type == ChangeTypes.Removed:
```
Comment (blinchk) [inline models/change.py:17]: **Templates with line-break in the code is great.**

**Review by blinchk — COMMENTED** [review]
Need some changes after my remarks.

**Comment by blinchk** [comment]
@Stepka1709, can you give some code review for this code?

File: main.py (hunk 12 above)
Comment (blinchk) [reply main.py:4]: Token should be defined in your IDE and environment variables at Docker.

**Review by stepka1709 — COMMENTED** [review]
...

**Review by blinchk — COMMENTED** [review]
#

**Comment by stepka1709** [comment]
Hi.

**Comment by blinchk** [comment]
This is a trash.

**Pull request closed without merging**